pvlib/ivtools/utility.py

```diff
@@ -0,0 +1,433 @@
+"""
+The ``pvlib.ivtools.utility.py`` module contains utility functions for fitting
+equations to IV curve data.
+
+"""
+
+import numpy as np
+import pandas as pd
+from collections import OrderedDict
+
+
+# A small number used to decide when a slope is equivalent to zero
+EPS = np.finfo('float').eps**(1/3)
+
+
+constants = OrderedDict()
+constants['E0'] = 1000.0
+constants['T0'] = 25.0
+constants['k'] = 1.38066e-23
+constants['q'] = 1.60218e-19
+
+
+def numdiff(x, f):
+    """
+    Compute first and second order derivative using possibly unequally
+    spaced data.
```

Comment:
Should compare this implementation to MetPy's

https://github.com/Unidata/MetPy/blob/931c1143379b637a3747b2c3c372568498630155/src/metpy/calc/tools.py#L986

https://github.com/Unidata/MetPy/blob/931c1143379b637a3747b2c3c372568498630155/src/metpy/calc/tools.py#L1076

Reply:
MetPy implements 3-point formulae for n-dimensional data, where the pvlib function uses 5-point formulae only for a function of a single variable. I don't see a way for pvlib to use the MetPy function, because the polynomial expressions for the weight at each point don't extend such that a 5point function can use a 3point function in it's calculation. Way back when I was working out this method there was a reason to use 5-point formulation, probably for some level of smoothing of the input data, but I don't recall exactly.

The MetPy function implement some logic to handle the partial windows at each end of the data, where the pvlib function ducks the issue and returns Nan. That could be an improvement to bring from MetPy to pvlib. I don't think it would substantially improve the downstream function that currently uses the output of this `numdiff` function.

I'm sure the pvlib function coding could be improved - it's very explicit. The MetPy function has sophistication that isn't needed for a function of a single variable.

Reply:
Hi @cwhanse, @wholmgren , & others,

Not a blocker to me, but curious if you had considered other numerical differentiation tools like:
* [`scipy.misc.derivative`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.misc.derivative.html#scipy.misc.derivative) - allows you to change the number of points used (from 3 to 5, etc) as long as it's odd - check the [source](https://github.com/scipy/scipy/blob/v1.5.2/scipy/misc/common.py#L75-L145) to see the weights used
* [`statsmodels.tools.numdiff`](https://www.statsmodels.org/stable/tools.html#numerical-differentiation) - also does complex numbers and can return the gradient, Jacobians, and Hessians
* [`numdifftools`](https://numdifftools.readthedocs.io/en/latest/) - latest release was June, 2019, but this is based on the [MATLAB derivest](https://www.mathworks.com/matlabcentral/fileexchange/13490-adaptive-robust-numerical-differentiation) originally authored by legendary Kodak mathematician [John D'errico](https://www.mathworks.com/matlabcentral/profile/authors/869215)
* [algopy](https://pythonhosted.org/algopy/) - I believe there used to be more packages like this, but Algopy seems to be the only one (in Python) that has survived. It requires significant effort, and overloads NumPy, but I believe it calculates derivatives of each operation as it occurs rather than of functional outputs

Anyway, just putting this out there, I don't actually expect you (or MetPy) to change anything, just always curious how or why there are so many tools that seem to all do the same thing, in apparent contradiction to [pep20: the Zen of Python](https://www.python.org/dev/peps/pep-0020/):
>There should be one-- and preferably only one --obvious way to do it.

Reply:
Thanks @mikofski . I would much prefer to import this capability from numpy or scipy than host it in pvlib. Maybe it's out there in python-land, when we wrote the Matlab code we couldn't find it. Looking at the resources you identified:

- scipy.misc.derivative requires equally spaced x-data, which is not a safe assumption for measured IV curves.
- statsmodels.tools.numdiff uses hardcoded 2 point (forward difference) or 3 point (centered difference) formula, although  it looks like it might accommodate unequal spacing, not clear to me.
- numdifftools looks like the primary use case is automated differentiation, which is related but different than calculating a numerical derivative from data which is the need for the sdm.fit_xxx_sandia functions.

MetPy's numdiff is based on the same reference I used, it could be made more general by adding higher order formulas, then imported here.

Reply:
Thanks for checking Cliff. Let's go with this implementation. 

Do we really want to make this part of the public API? Same concern for the qspline function.